### DISCUSSION TITLE
Tips for domain randomization for training legged locomotion

### DISCUSSION THREAD

**Discussion by TRo202** [isaac-sim/IsaacLab#2813] [Q&A]
Hi. I’m currently working on legged locomotion, and have sim-to-real issues. Training works fine and fast with my algorithm and settings for reward gains, hyperparameters, and etc. However, if I apply noise in observation and domain randomization, training doesn’t work well. I wonder how other people in robotics field deal with domain randomization.

Do you apply noise and domain randomization from the beginning right away or first train without them, then load the pre-trained model and start training again with noise and domain randomization?

Should I try to reduce the noise and randomization range or try to tune the hyperparameters? I honestly doubt if reducing the noise and randomization range is fine since then the trained model will be less robust to sim-to-real gaps.

**Comment by RandomOakForest** [comment]
Thank you for starting this discussion. Here is a summary that is still under review, but that you may find helpful. 

In legged locomotion, **domain randomization** and **observation noise** are critical for bridging the sim-to-real gap, but their application requires careful strategy to avoid destabilizing training.

**Best Practices for Applying Domain Randomization and Noise**

**1. Progressive Curriculum:**

- **Do not apply strong noise and domain randomization from the very start.**
    - Begin training *without* or with *minimal* randomization and noise.
    - Once the policy demonstrates stable learning and good performance in the nominal (clean) environment, gradually introduce noise and randomization.
    - This staged approach is called *curriculum domain randomization* and is widely used to prevent early learning collapse and ensure policy robustness[^1][^2].

**2. Gradual Increase of Randomization:**

- Start with **narrow/randomization ranges** (e.g., friction, mass, sensor noise) and incrementally widen them as training progresses and the policy improves[^1][^2].
- Automatic Domain Randomization (ADR) techniques can automate this process by expanding parameter ranges only when the policy can maintain a certain performance threshold, providing a natural curriculum and improved robustness[^2].

**3. Hyperparameter Tuning:**

- Introducing noise and randomization often requires **retuning RL hyperparameters** (e.g., learning rate, batch size, reward scaling) because the learning dynamics change.
- If training becomes unstable after adding randomization, first try adjusting these hyperparameters before reducing the randomization ranges[^1].

**4. Parameter Range Selection:**

- Use **system identification** or real-world measurements to set realistic randomization bounds (e.g., measure friction on your lab floor to set friction randomization limits)[^1][^3].
- Excessively wide ranges can destabilize learning and may lead to policies that are too conservative or fail to converge[^1].

**5. Implementation in Isaac Lab:**

- In Isaac Lab, domain randomization is configured via a class specifying which parameters to randomize, their ranges, and the schedule (e.g., at reset or at intervals)[^4].
- Example configuration (from Isaac Lab docs):

```python
@configclass
class EventCfg:
    robot_physics_material = EventTerm(
        func=mdp.randomize_rigid_body_material,
        mode="reset",
        params={
            "asset_cfg": SceneEntityCfg("robot", body_names=".*"),
            "static_friction_range": (0.7, 1.3),
            "dynamic_friction_range": (1.0, 1.0),
            "restitution_range": (1.0, 1.0),
            "num_buckets": 250,
        },
    )
    robot_joint_stiffness_and_damping = EventTerm(
        func=mdp.randomize_actuator_gains,
        mode="reset",
        params={
            "asset_cfg": SceneEntityCfg("robot", joint_names=".*"),
            "stiffness_distribution_params": (0.75, 1.5),
            "damping_distribution_params": (0.3, 3.0),
            "operation": "scale",
            "distribution": "log_uniform",
        },
    )
```

- This allows you to randomize robot friction, joint gains, and other properties at episode resets or at specified intervals[^4].

**6. Types of Randomization:**

- Randomize parameters that most affect sim-to-real transfer:
    - **Physical properties:** friction, mass, restitution, actuator gains[^3][^1][^4]
    - **Sensor noise:** add observation noise matching real sensor characteristics[^1][^4]
    - **Environment dynamics:** gravity, terrain roughness, initial robot state, external perturbations[^3][^4]

**Summary Table: Curriculum vs. Direct Randomization**


| Approach | When to Use | Pros | Cons |
| :-- | :-- | :-- | :-- |
| Curriculum (Progressive) | Most common, especially for complex tasks | Stable learning, better convergence | Requires schedule design |
| Direct (Full from start) | Simple tasks, robust policies | Simpler setup | High risk of early training collapse |

**Key Recommendations:**

- **Start with curriculum randomization:** Train first in a clean environment, then progressively add noise and randomization.
- **Tune hyperparameters** after introducing randomization.
- **Do not reduce randomization range** just to make training easier; instead, use curriculum and tuning to reach robustness.
- **Measure real-world parameters** to set realistic randomization bounds.


[^1]: https://app.studyraid.com/en/read/15451/536759/domain-randomization-to-handle-reality-gap

[^2]: https://github.com/isaac-sim/IsaacGymEnvs/blob/main/docs/domain_randomization.md

[^3]: https://arxiv.org/html/2502.10894v1

[^4]: https://isaac-sim.github.io/IsaacLab/main/source/migration/migrating_from_omniisaacgymenvs.html

[^5]: https://docs.isaacsim.omniverse.nvidia.com/4.5.0/py/source/extensions/isaacsim.replicator.domain_randomization/docs/index.html

[^6]: https://github.com/isaac-sim/IsaacLab/discussions/2813

[^7]: https://isaac-sim.github.io/IsaacLab/

[^8]: https://docs.isaacsim.omniverse.nvidia.com/latest/robot_setup_tutorials/tutorial_rig_legged_robot.html

[^9]: https://developer.nvidia.com/blog/closing-the-sim-to-real-gap-training-spot-quadruped-locomotion-with-nvidia-isaac-lab/

[^10]: https://docs.isaacsim.omniverse.nvidia.com/latest/isaac_lab_tutorials/tutorial_policy_deployment.html

[^11]: https://www.youtube.com/watch?v=iBB7D8r_I1g

[^12]: https://www.youtube.com/watch?v=LuInGAfkpkY

[^13]: https://forums.developer.nvidia.com/t/tips-for-domain-randomization-for-training-legged-locomotion/337002

[^14]: https://github.com/leggedrobotics/legged_gym

[^15]: https://forums.developer.nvidia.com/t/the-issue-of-abnormal-joint-movement-in-the-legged-robot-in-isaac-lab/317165

[^16]: https://github.com/robotlearning123/awesome-isaac-gym/blob/main/README.md

[^17]: https://arxiv.org/html/2503.01255v1

[^18]: https://arxiv.org/pdf/2304.13150.pdf

[^19]: https://forums.developer.nvidia.com/t/dynamic-randomization-with-center-of-mass-and-inertia/288441

[^20]: https://github.com/isaac-sim/OmniIsaacGymEnvs/blob/main/docs/framework/domain_randomization.md

[^21]: https://junfeng-long.github.io/HIMLoco/static/pdfs/HIMLocomotion.pdf

[^22]: https://uan.csail.mit.edu/rsc/paper.pdf

[^23]: https://arxiv.org/pdf/1903.11774.pdf

[^24]: https://publications.idiap.ch/attachments/papers/2024/Xue_CORL_2024.pdf

[^25]: https://rai-inst.com/wp-content/uploads/2025/05/ICRA_2025__High_Performance_Reinforcement_Learning_on_Spot__Optimizing_Simulation_Parameters_with_Distributional_Measures.pdf

[^26]: https://openreview.net/forum?id=9JQXuyzdGL
